make_prg/utils/seq_utils.py

```diff
@@ -0,0 +1,220 @@
+    for record in alignment:
+        yield str(record.seq)
+
+
+def get_number_of_unique_ungapped_sequences(sub_alignment: MSA) -> int:
+    alignment_seqs = get_alignment_seqs(sub_alignment)
+    ungapped_sequences = map(ungap, alignment_seqs)
+    deduplicated_ungapped_sequences = remove_duplicates(ungapped_sequences)
+    number_of_unique_nongapped_sequences = count(deduplicated_ungapped_sequences)
+    return number_of_unique_nongapped_sequences
+
+
+def get_number_of_unique_gapped_sequences(sub_alignment: MSA) -> int:
+    alignment_seqs = get_alignment_seqs(sub_alignment)
+    deduplicated_gapped_sequences = remove_duplicates(alignment_seqs)
+    number_of_unique_gapped_sequences = count(deduplicated_gapped_sequences)
+    return number_of_unique_gapped_sequences
+
+
+class SequenceCurationError(Exception):
+    pass
+
+
+class SequenceExpander:
+    iupac = {
+        "R": "GA",
+        "Y": "TC",
+        "K": "GT",
+        "M": "AC",
+        "S": "GC",
+        "W": "AT",
+        "A": "A",
+        "C": "C",
+        "G": "G",
+        "T": "T",
+    }
+    expandable_bases = set(iupac.keys())
+    allowed_bases = expandable_bases.union({"N"})
+    standard_bases = {"A", "C", "G", "T"}
+    ambiguous_bases = expandable_bases.difference(standard_bases)
+
+    @classmethod
+    def check_if_there_is_sequence_with_disallowed_bases(cls, sequences: List[str]):
+        for sequence in sequences:
+            if not set(sequence).issubset(cls.allowed_bases):
+                raise SequenceCurationError(
+                    "A slice of a sequence has a disallowed base.\n"
+                    f"Allowed bases: {cls.allowed_bases}.\n"
+                    f"Sequence: {sequence}\n"
+                    f"Redo sequence curation.\n"
+                )
+
+    @classmethod
+    def check_all_sequences_are_composed_of_ACGT(cls, sequences: List[str]):
+        for sequence in sequences:
+            sequence_is_composed_of_ACGT_only = set(sequence).issubset(cls.standard_bases)
+            assert sequence_is_composed_of_ACGT_only, f"Sequence ({sequence}) should be composed of ACTG only."
+
+    @classmethod
+    def get_expanded_sequences(cls, sequences: List[str]) -> Sequences:
+        """
+        Expand sequences in the given list of sequences, following the translation table in SequenceExpander.iupac.
+        It does the following steps:
+            1. Check that we don't have disallowed bases;
+            2. Remove sequences with N;
+            3. Duplicate sequences containing RYKMSW, replacing with AGCT alternatives;
+        Note 1: The sequences are deliberately returned in the order they are received.
+        Note 2: Returned sequences are composed of ACGT only
+        """
+
+        cls.check_if_there_is_sequence_with_disallowed_bases(sequences)
+
+        expanded_seqs = []
+        expanded_set = set()
+        deduplicated_gapless_seqs = remove_duplicates(sequences)
+        for seq in deduplicated_gapless_seqs:
```

Comment:
I find it clearer to just do `for seq in remove_duplicates(sequences)`- but this is a minute point, can ignore

Reply:
although because `remove_duplicates` returns a generator i think it's best not to assign a variable to it- the generator will be exhausted after traversal so the variable becomes useless

Reply:
I do agree, leaving an exhausted generator in the code is error prone, implementing your suggestion

Reply:
Done in https://github.com/iqbal-lab-org/make_prg/pull/34/commits/f3bfa155a595c4a8f864d3d74abcf654cdf216b0